Refactor genomic range sharing into utils-java.

deflaux opened the pull request:
Logic to shard genomic ranges should be all in one place to be reused by both dataflow and spark.  Its also important to ensure that the full collection of shards is shuffled so that parallel requests will be distributed across the dataset.

This pull request is part 1 of many parts to refactor some logic currently in dataflow-java down into utils-java so that it can be more widely re-used.  Since this code may change yet again, sending these pull requests to a branch for now instead of master.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -75,23 +85,6 @@ public String toString() {
-    return getShards(DEFAULT_NUMBER_OF_BASES_PER_SHARD);
-  }
-
   public SearchVariantsRequest getVariantsRequest(String variantSetId) {
```
In my opinion this method belongs in some variant-related class, not here. It would be something like an interface that's implemented by a VCF file reader and a Genomics API client.

The benefit is that (i) others can add other implementations if they want (reading from GCS, Hadoop, whatever comes next), and (ii) the contig class changes less (this is good because lots of code will come to depend on contig).

deflaux replied:
Agreed.  This will happen in a separate PR.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -100,6 +93,15 @@ public SearchVariantsRequest getVariantsRequest(String variantSetId) {
         .setEnd(end);
   }
 
+  public StreamVariantsRequest getStreamVariantsRequest(String variantSetId) {
```
So instead of adding a new method here, I'd move the original method and place this one along with it.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -108,7 +110,41 @@ public SearchReadsRequest getReadsRequest(String readGroupSetId) {
   }
 
-  public static Iterable<Contig> parseContigsFromCommandLine(String contigsArgument) {
+  public StreamReadsRequest getStreamReadsRequest(String readGroupSetId) {
```
Same thing for those, though I understand if you want to do that in a separate PR.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -108,7 +110,41 @@ public SearchReadsRequest getReadsRequest(String readGroupSetId) {
+        .build();
+  }
+
+  private List<Contig> getShards() {
```
As a matter of taste, I much prefer when all the public methods are first in the file, then all the private ones at the bottom. The reason is that most people are just users of the class, so they can stop reading after the last public method.

deflaux replied:
Will do!

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
   }
+  
+  /**
+   * Get a list of Contigs representing sharded windows of all genomic regions within the variant set.
```
We're using 4 different words to mean the same thing: contig, shard, region, and interval. Let's try to focus on just one. Unless we want to rename this class, I suggest that our best candidate is "contig". We'll use that to mean "a contiguous region of the genome". 

So this function can be described as (i) loading the specified variants via the API, (ii) casting those variants to contigs - i.e. only keeping the information about where the variant is.

I'm a bit confused about why this should be here. Loading variants via the API has nothing to do with contigs so it doesn't feel like it belongs here. The second part sounds like it should be a method on the variant object: variant.getContig(). Surely that either already exists, or we should add it because knowing where a variant is seems fundamental.

Given this, I don't see why either step belongs here. Dataflow is good at combining DoFns, so it's already easy enough to write variantIDs.apply(new LoadVariantFn()).apply(new CastToContigFn()). No?

deflaux replied:
This doesn't load the variants.

It just determines what contigs a variant set has and then also shards them.  It comes into play when you use command line argument `--allReferences` via spark or dataflow.

jean-philippe-martin replied:
You're right, I was confused. This also explains why we spit out contigs that start at 0 - that's because it's what the API gives us. I still would like to see all those Genomic API-dependent features out of the contig class and into some helper class, but I'll grant you this is a matter of taste.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
+   * @return a shuffled list of shards
+   * @throws IOException
+   */
+  public static List<Contig> getAllShardsInVariantSet(final Genomics genomics, final String variantSetId,
```
I would not include this form of the function. It's not needed in a minimal API, and adds work for us. Same for all other variants that pick DEFAULT_NUMBER_OF_BASES_PER_SHARD instead of getting it via argument.

deflaux replied:
Will remove.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
+  }
+  
+  /**
+   * Get a list of Contigs representing sharded windows of all genomic regions within the variant set.
```
This doesn't tell the whole story - unless I'm mistaken it looks like those contigs start at 0, instead of starting at the variant's start position (line 182). So this isn't actually a simple typecast from Variant to bounding interval.

deflaux replied:
variantsets.get for v1beta2 only returns the upper bound per reference, hence the hard-coded lower bound

jean-philippe-martin replied:
Indeed, so the comment should reflect that reality.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
+   *                            format is chromosome:start:end[,chromosome:start:end]
+   * @return a shuffled list of shards
+   */
+  public static List<Contig> getSpecifiedShards(final String contigsArgument) {
```
Let's agree on a single word to mean a genomic interval. I think that word is contig. So this function might be renamed e.g. to "parse", as in "Contig.parse("chr1:1:100")". Or "of", a name Dataflow often uses for static constructor-like methods.

deflaux replied:
Will remove.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Contig.java:
```diff
@@ -148,5 +184,78 @@ public Contig apply(String contigString) {
+   * @param numberOfBasesPerShard - the maximum size of the shard window in terms of number of bases
+   * @return a shuffled list of shards
+   */
+  public static List<Contig> getSpecifiedShards(final String contigsArgument, final long numberOfBasesPerShard) {
```
Let's agree on a single word to mean a genomic interval. I think that word is contig. So this function might be renamed e.g. to "parse", as in "Contig.parse("chr1:1:100")". Or "of", a name Dataflow often uses for static constructor-like methods.

deflaux replied:
Agreed on using fewer synonyms.  There are times though that we have both contigs and shards.  Take a look at the example below and let me know what you think?

* https://github.com/deflaux/dataflow-java/blob/refactor-part1/src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java#L182
* https://github.com/deflaux/dataflow-java/blob/refactor-part1/src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java#L195 the sharding happens in a different class

jean-philippe-martin replied:
Agreed, that code uses both terms. I'm not asking you to go change it, but having seen those terms a few times I wanted to start the conversation - if we reach an agreement then the big rename can be a separate CL.

jean-philippe-martin commented on src/test/java/com/google/cloud/genomics/utils/ContigTest.java:
```diff
@@ -15,43 +15,76 @@
 
 @RunWith(JUnit4.class)
 public class ContigTest {
+  
```
I see you're not testing all of the new functions. It would be nice if we did.

deflaux replied:
The new methods that copy Contig fields into request fields aren't that interesting to test.

The test for the new method `getSpecifiedShards` exercises it and the private methods upon which it depends.

There is no test for the new method `getAllShardsInVariantSet` because it just makes an api call and then exercises the same logic as is used by `getSpecifiedShards`.  I could mock it out or add an integration test.

deflaux commented:
@jean-philippe-martin PTAL

* rearranged code per feedback and added integration tests
* also refactored read sharding from dataflow-java to utils-java
* please be sure to view https://github.com/googlegenomics/dataflow-java/pull/116 so that you can see where much of this code came from

Happy to rename anything to clean up/consolidate terminology.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java:
```diff
@@ -0,0 +1,89 @@
+  
+  /**
+   * Gets ReadGroupSetIds from a given datasetId using the Genomics API.
+   */
```
perhaps add: "Throws an exception if there is none", since this behavior may be surprising to some (reading from files or console doesn't throw if the input is empty).

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/GenomicsUtils.java:
```diff
@@ -0,0 +1,89 @@
+  }
+  
+  /**
+   * Gets VariantSetIds from a given datasetId using the Genomics API.
```
mention the exception here, and in the next method too.

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/Paginator.java:
```diff
@@ -94,9 +94,9 @@
+ *    searchReadsets.search(new SearchReadsetsRequest().setDatasetId(datasetId))) &#123;
  *  // do something with readset
- *}
+ *&#125;
```
did we mean to do this? Is it necessary for something? It certainly makes the comment harder to read in the IDE.

deflaux replied:
Yes, it was to clear an eclipse error on javadoc formatting.

jean-philippe-martin replied:
I checked out [stackoverflow](http://stackoverflow.com/questions/647195/how-do-you-escape-curly-braces-in-javadoc-inline-tags-such-as-the-code-tag) for a bit of background information, and man is this messed up.
 * `<code>` requires you to escape angle brackets
 * `{@code ... }` requires you to escape braces

jean-philippe-martin commented on src/main/java/com/google/cloud/genomics/utils/ShardUtils.java:
```diff
@@ -0,0 +1,297 @@
+ * 
+ * For Dataflow, custom sources should be preferred over this manual sharding approach.
+ *
+ */
```
excellent comment!

jean-philippe-martin commented on src/test/java/com/google/cloud/genomics/utils/IntegrationTestHelper.java:
```diff
@@ -0,0 +1,108 @@
+  public final String API_KEY = System.getenv("GOOGLE_API_KEY");
+  public final OfflineAuth auth;
+  
+  public IntegrationTestHelper() throws IOException, GeneralSecurityException {
```
I like the idea of a test helper. Here there's a risk that users may forget to call the constructor - all methods are static after all, and it's all too easy to write IntegreationTestHelper.API_KEY.

What I'd like to suggest instead is to have either
(a) API_KEY and auth are protected, only accessible via non-static getters. Thus it's impossible to forget to call the constructor (if we need it)
(b) (I prefer this one) API_KEY and auth are static protected, only accessible via static getters. The getters will grab the information on the first call, stash it for later calls. The upside is that if we don't need auth but need something else from the helper, we can still use it without getting an error. The downside is this isn't thread-safe so we either live with that or spend a bit more time making it so.

jean-philippe-martin commented on src/test/java/com/google/cloud/genomics/utils/ShardUtilsITCase.java:
```diff
@@ -0,0 +1,111 @@
+    };
+    
+    // These shards are "too big" to use in practice but for this test it keeps the
+    // expected result from getting crazy long.
```
Good call :)

jean-philippe-martin commented:
Thank you! LGTM, aside from minor comments. Feel free to merge once ready.

deflaux commented:
Javadocs added.  Will update `IntegrationTestHelper.API_KEY` in both utils-java and dataflow-java in the next set of pull requests.
